### PR TITLE
Implement TruncatedDistribution

### PR DESCRIPTION
**Pull request by alicanb**
this PR adds:
- `.cdf()` and `icdf()` methods for `Distribution` and tests. (Populated only for `Normal` for now)
- `TruncatedDistribution` class
- `TruncatedNormal` class

### CODE REVIEW

**Comment by vishwakftw** [comment]
Will the `.cdf` methods for existing distributions be added in this PR, or a new PR altogether?

**Comment by alicanb** [comment]
I just added minimal working example (+tests) to get TruncatedNormal working. With cdfs for all current distributions I think this PR would be too large. What do you think?

**Comment by vishwakftw** [comment]
I thought the same too. Maybe after this is merged, I can start working on the populating PR. Hope that is fine.

**Comment by alicanb** [comment]
You can cherry-pick the first commit to your branch and start working in parallel if you want?

**Comment by tbrx** [comment]
I was actually just thinking about this today and was exploring how disasterous it would be to try to implement a "generic" TruncatedDistribution, where we use inverse transform sampling to generate from it.

Some plots in this gist: https://gist.github.com/tbrx/18e7579d9b7ff7c2a84c17c300555fc1

Basically, it's pretty bad numerically once you are more than four standard deviations away from the mean, on a Gaussian, and falls apart entirely a little past five. This doesn't give me high hopes for e.g. Gamma…

I looked at the Scipy code this morning, and it actually appears to use inverse transform sampling for truncated normals. Higher-precision floating point though means that they can get quite far away from the mean before this is an issue.

File: test/test_distributions.py [hunk 1]
```diff
@@ -2309,6 +2309,28 @@ def test_variance_stddev(self):
                 self.assertEqual(pytorch_dist.variance, scipy_dist.var(), allow_inf=True, message=pytorch_dist)
                 self.assertEqual(pytorch_dist.stddev, scipy_dist.var() ** 0.5, message=pytorch_dist)
 
+    def test_cdf(self):
+        set_rng_seed(0)  # see Note [Randomized statistical tests]
+        for pytorch_dist, scipy_dist in self.distribution_pairs:
+            samples = pytorch_dist.sample((5,))
+            try:
```
Comment (fritzo) [inline test/test_distributions.py]: It's safest to enclose as little as possible in a try-except:
```py
try:
    pytorch_cdf = pytorch_dist.cdf(samples)
except NotImplementedError:
    pass
self.assertEqual(pytorch_cdf, scipy_dist.cdf(samples), message=pytorch_dist)
```

File: test/test_distributions.py [hunk 2]
```diff
@@ -2309,6 +2309,28 @@ def test_variance_stddev(self):
...
+                                 scipy_dist.cdf(samples),
+                                 message=pytorch_dist)
+            except NotImplementedError:
+                pass
+
+    def test_icdf(self):
+        set_rng_seed(0)  # see Note [Randomized statistical tests]
+        for pytorch_dist, scipy_dist in self.distribution_pairs:
+            samples = Variable(torch.rand((5,) + pytorch_dist.batch_shape))
+            try:
```
Comment (fritzo) [inline test/test_distributions.py]: ditto, enclose as little as possible

File: torch/distributions/truncated_distribution.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+class TruncatedDistribution(Distribution):
+    r"""
+    Extension of the Distribution class, which applies truncation to base distribution.
+    """
+    def __init__(self, base_distribution, lower_bound=-float('inf'), upper_bound=float('inf'), *args, **kwargs):
+        super(TruncatedDistribution, self).__init__(*args, **kwargs)
+        self.base_dist = base_distribution
+        self.lower_bound, self.upper_bound, _ = broadcast_all(lower_bound, upper_bound,
+                                                              getattr(self.base_dist,
```
Comment (fritzo) [inline torch/distributions/truncated_distribution.py]: This looks really dangerous. Why do we need to broadcast? Can we simply set
```py
self.lower_bound = lower_bound
self.upper_bound = upper_bound
```

File: torch/distributions/truncated_distribution.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        return self.base_dist.event_shape
+
+    def rsample(self, sample_shape=torch.Size()):
+        """
+        Generates a sample_shape shaped sample or sample_shape shaped batch of
+        samples if the distribution parameters are batched via inverse cdf sampling. Note that this
+        is a generic sampler which is not the most efficient or accurate around tails of base distribution.
+        """
+        shape = shape = self._extended_shape(sample_shape)
+        u = getattr(self.base_dist, list(self.base_dist.params.keys())[0]).new(shape).uniform_()
```
Comment (fritzo) [inline torch/distributions/truncated_distribution.py]: This looks dangerous. I wish we had a `.new()` method to create a correctly-placed tensor from given distribution.

@apaszke Is there an established pattern to do this? Can we define a `.new_tensor()` method or something? This has been coming up often. Some of our distributions define a private `_new()` but we haven't exposed this as a general interface.

Comment (fritzo) [reply torch/distributions/truncated_distribution.py]: It seems simple and safe to define a method-as-property like
```py
class Distribution(object):
    @property
    def new_tensor(self):
        raise NotImplementedError

class Normal(Distribution):
    @property
    def new_tensor(self):
        return self.loc.new
```

File: torch/distributions/truncated_distribution.py (hunk 3 above)
Comment (alicanb) [reply torch/distributions/truncated_distribution.py]: I was thinking about supporting batched bounds while writing that part, but I gave up that idea later & forgot to change it.

**Comment by fritzo** [comment]
Re: numerical stability, one option would be to use rejection sampling to draw samples and merely use the cdf derivative to compute reparameterized gradients:
```py
def sample(self):
    ...use rejection sampling...

def rsample(self):
    x = self.sample()  # detached
    cdf = self.cdf(x)
    pdf = self.log_prob(x).exp()
    return x + (cdf.detach() - cdf) / pdf.detach()
```

**Comment by alicanb** [comment]
Do you think rejection sampling would be fast enough? Should we write low
level functions for it?

**Comment by fritzo** [comment]
> Do you think rejection sampling would be fast enough? 

It would be cheap if we rejection sampled when `cdf(lower_bound) < 0.02` use inverse cdf otherwise. This would require four branches: rejection-sample + inverse-cdf for each of the lower_bound and upper_bound. We would not need any low level code; see e.g. [pyro.distributions.Rejector.sample()](https://github.com/uber/pyro/blob/de4a1cb/pyro/distributions/rejector.py#L42).

**Comment by alicanb** [comment]
@fritzo thanks, I take a look try to come up with something.

**Comment by fritzo** [comment]
I think `.cdf()` methods have many uses outside of `TruncatedDistribution`, and are worth adding even if this `TruncatedDistribution` ends up being too tricky to add in PyTorch 0.4

**Comment by tbrx** [comment]
Oh, I completely agree that adding `.cdf` and its inverse are useful independently of truncated distributions! I'm just wary of releasing a "general" `TruncatedDistribution` without some caveats, or issuing warnings about lack of precision.

For the truncated normal, example it seems like there are only 58 distinct floating point values between 4.5 and infinity. It seems like if your bounds are within ±4 standard deviations though this would work pretty much fine! Maybe that is the more common case than sampling or evaluating tail probabilities anyway.

**Comment by alicanb** [comment]
Here are 2 gists for `TruncatedNormal`, one with [`lower_bound`=3](https://gist.github.com/alicanb/efae77cfec49dd42d23840ce55b418dd) and one with [`lower_bound`=4](https://gist.github.com/alicanb/2ca8df38a59b1ffd1d4e25e364378f3b). Something doesn't add up with rejection sampling, @fritzo @tbrx take a look, what is a proper way to test this stuff?

**Comment by fritzo** [comment]
Do I understand correctly that the difficult case is when you're truncating e.g. a `Normal(0, 1)` to its tail like `[10, float('inf'))`? If inverse-cdf doesn't work here, then I don't know what will.

**Comment by alicanb** [comment]
Here is an [updated gist](https://gist.github.com/alicanb/c9e6567b7c512140ed43916b4dd30106). Sampling from 4.5 sigma looks problematic, but [sampling from 4 sigma looks okish](https://gist.github.com/alicanb/c9e6567b7c512140ed43916b4dd30106/a21132a2cebdeff93947cd7b83f538dda3174ce0)

**Comment by tbrx** [comment]
I don't think inverse CDF sampling will work very well for a Normal(0,1) outside of the region [-4, 4] or maybe [-4.5, 4.5] in a pinch…

There are algorithms for sampling from the tail of a gaussian (e.g. on [4, \infty) ) in chapter 9 of http://www.nrbook.com/devroye/. This doesn't help, though, with computing the `.log_prob` once we get > 5 or so. Sorry I've been busy with a deadline — will try to look at this closer over the weekend or on Monday.

**Comment by alicanb** [comment]
I implemented sampling from tail algorithm, it's fast and looks good! Here's a [gist](https://gist.github.com/alicanb/868fdbaeb9b3ee98737fdf47a2ada58d).

**Comment by tbrx** [comment]
This is cool!! That would work really well for `.sample`, at least. For `.rsample` maybe we can handle a loss of precision.

I was wondering if there was a way of maybe directly approximating `Phi(x) - Phi(4.0)`, or something like that, to help compute the denominator…?

**Comment by ragulpr** [comment]
I'm very interested in this, I'm working on a similar thing. I'm calling it ConditionalExcessDistribution, but I'm really only looking at right censored (truncated) things at the moment.

**Comment by alicanb** [comment]
New gist time 😄 This time I also have sampling times. https://gist.github.com/alicanb/c9e6567b7c512140ed43916b4dd30106 . At this point I'm inclined towards having `TruncatedDistribution` only have inverse-cdf sampling implemented, and `TruncatedNormal` using [robert's algorithm](https://arxiv.org/abs/0907.4010). Still working on the precision problem with `erf`...

**Comment by fritzo** [comment]
> I'm inclined towards having `TruncatedDistribution` only have inverse-cdf sampling implemented, and `TruncatedNormal` using robert's algorithm

That sounds reasonable, implementing one new generic distribution and one specific special-case distribution. It even makes sense to send them in the same PR.

File: torch/distributions/truncated_distribution.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    r"""
+    Extension of the Distribution class, which applies truncation to base distribution.
+    """
+    def __init__(self, base_distribution, lower_bound=-float('inf'), upper_bound=float('inf'), sample_method='inversion', *args, **kwargs):
+        super(TruncatedDistribution, self).__init__(*args, **kwargs)
+        self.lower_bound, self.upper_bound = broadcast_all(lower_bound, upper_bound)
+        self.base_dist = base_distribution
+        cdf_low, cdf_high = self.base_dist.cdf(self.lower_bound), self.base_dist.cdf(self.upper_bound)
+        if sample_method in ['rejection', 'inversion']:
+            self.sample = {'rejection': self._rejection_sample, 'inversion': self._inversion_sample}[sample_method]
```
Comment (fritzo) [inline torch/distributions/truncated_distribution.py]: This creates a circular reference and leaks memory.

File: torch/distributions/truncated_distribution.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+    @property
+    def batch_shape(self):
+        return self.base_dist.batch_shape
+
+    @property
+    def event_shape(self):
+        return self.base_dist.event_shape
+
+    def _inversion_sample(self, sample_shape=torch.Size()):
```
Comment (fritzo) [inline torch/distributions/truncated_distribution.py]: I'm inclined to implement `Inversion` and `Rejection` as different classes because the interfaces differ: `Inversion` allows reparametrization hence allows an `.rsample()` whereas `Rejection` is not reparametrizable and hence only implements `.sample()` (It can be  partially reparametrized via RSVI but that requires yet a different interface). Also, Pyro defines a different `Rejector` class to do rejection sampling given a more general rejection criterion.

File: torch/distributions/truncated_normal.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        [torch.FloatTensor of size 1]
+
+    Args:
+        loc (float or Tensor or Variable): mean of log of distribution
+        scale (float or Tensor or Variable): standard deviation of log ofthe distribution
+    """
+    def __init__(self, loc, scale, lower_bound=-float('inf'), upper_bound=float('inf'), sample_method='robert', *args, **kwargs):
+        super(TruncatedNormal, self).__init__(Normal(loc, scale), lower_bound, upper_bound, *args, **kwargs)
+        if sample_method in {'exp', 'robert'}:
+            self.sample = {'exp':self._exp_proposal, 'robert': self._robert_sample}[sample_method]
```
Comment (fritzo) [inline torch/distributions/truncated_normal.py]: This creates a circular reference. It's better to simply define an `if` statement in an `.rsample()` method.

File: torch/distributions/truncated_distribution.py (hunk 6 above)
Comment (alicanb) [reply torch/distributions/truncated_distribution.py]: I'm inclined to omit rejection sampling from pytorch actually. It's hard to make it work efficiently oob for a range of distributions. What do you think?

Comment (fritzo) [reply torch/distributions/truncated_distribution.py]: I agree, let's omit rejection sampling.
